lib/PromoteStackPointer/ComputeStackAccessesBoundsPass.cpp

```diff
@@ -2,58 +2,188 @@
 // This file is distributed under the MIT License. See LICENSE.md for details.
 //
 
+#include <optional>
+#include <utility>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
 #include "llvm/Analysis/LazyValueInfo.h"
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/Analysis/PostDominators.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+#include "llvm/IR/BasicBlock.h"
 #include "llvm/IR/ConstantRange.h"
+#include "llvm/IR/Dominators.h"
 #include "llvm/IR/Instructions.h"
 
 #include "revng/Model/IRHelpers.h"
+#include "revng/Support/IRHelpers.h"
 
 #include "revng-c/PromoteStackPointer/ComputeStackAccessesBoundsPass.h"
 #include "revng-c/PromoteStackPointer/InstrumentStackAccessesPass.h"
 #include "revng-c/Support/FunctionTags.h"
 
 using namespace llvm;
 
-bool ComputeStackAccessesBoundsPass::runOnModule(Module &M) {
+static Logger<> Log("compute-stack-accesses-bounds");
+
+template<typename T, std::ranges::range R>
+bool areAll(const R &Range) {
+  return llvm::all_of(Range, [](auto *Object) { return isa<T>(Object); });
+}
+
+static ConstantRange
+getSCEVBoundaries(ScalarEvolution &SE, PostDominatorTree &PDT, Instruction *I) {
+  const auto *SCEV = SE.getSCEV(I);
+  auto FullSet = ConstantRange::getFull(SCEV->getType()->getIntegerBitWidth());
+
+  revng_log(Log, "getSCEVBoundaries on " << dumpToString(*SCEV));
+  LoggerIndent<> Indent(Log);
+
+  auto *AddRec = dyn_cast<SCEVAddRecExpr>(SCEV);
+  if (AddRec == nullptr)
+    return FullSet;
+
+  // Only consider loop in loop simplify form
+  auto *Loop = AddRec->getLoop();
+  if (not Loop->isLoopSimplifyForm())
+    return FullSet;
+
+  // Only provide a result if the block post-dominates the header, which
+  // implies the block sees one more value than the backedge taken count
+  BasicBlock *Location = I->getParent();
+  if (not PDT.dominates(Location, Loop->getHeader()))
+    return FullSet;
+
+  // Collect operands
+  // TODO: handle multiple AddRecs
+  const auto *Start = AddRec->getStart();
+  const class SCEV *End = nullptr;
+  auto *BackedgeTakenCount = SE.getBackedgeTakenCount(Loop);
+  const auto *Stride = AddRec->getStepRecurrence(SE);
+
+  if (not areAll<SCEVConstant>(std::vector{
+        Start, BackedgeTakenCount, Stride })) {
+    return FullSet;
+  }
+  bool Increasing = cast<SCEVConstant>(Stride)->getAPInt().isStrictlyPositive();
+  revng_log(Log, "Increasing: " << Increasing);
+
+  if (Increasing) {
+    // Compute end value
+    llvm::SmallVector<const llvm::SCEV *, 4> Addends = {
+      Start,
+      SE.getMulExpr(Stride, BackedgeTakenCount),
+      SE.getOne(Start->getType())
+    };
+    End = SE.getAddExpr(Addends);
+  } else {
+    End = SE.getAddExpr(Start, SE.getOne(Start->getType()));
+    Start = SE.getAddExpr(Start, SE.getMulExpr(Stride, BackedgeTakenCount));
+  }
```

Comment:
@pfez We now also handle backward loops, I added a test too. See `backward_memcpy_from_stack_arguments` in this PR and in the [revng-qa's one](https://github.com/revng/revng-qa/pull/28/files).

See if it looks good to you.

Reply:
Yeah great thanks!